models/utils/detect_face.py

```diff
@@ -5,13 +5,49 @@
 from PIL import Image
 import numpy as np
 import os
+import math
 
 # OpenCV is optional, but required if using numpy arrays instead of PIL
 try:
     import cv2
 except:
     pass
 
+def fixed_batch_processing(im_data, model, model_name):
+    """
+    Use fixed-batch size for rnet and onet to avoid out of GPU memory error
+    im_data:   input for 'rnet' or 'onet'
+    model : rnet or onet
+    model_name : 'rnet' or 'onet'
+    """
+    assert (model_name in ['rnet', 'onet'])
+
+    batch_size = 512
+    n = im_data.shape[0]
+    n_iter = int(math.ceil(n / float(batch_size)))
+    b = []
+    a = []
+    is_onet = (model_name == 'onet')
+    if is_onet:
+        c = []
+
+    for i in range(n_iter):
+        min_ind = i * batch_size
+        max_ind = (i + 1) * batch_size
+        batch = im_data[min_ind:max_ind]
+        out = model(batch)
+        if is_onet:
+            b.append(out[0])
+            c.append(out[1])
+            a.append(out[2])
+        else:
+            b.append(out[0])
+            a.append(out[1])
+
+    if is_onet:
+        return torch.cat(b, dim=0), torch.cat(c, dim=0), torch.cat(a, dim=0)
+    else:
+        return torch.cat(b, dim=0), torch.cat(a, dim=0)
```

Comment:
This entire function can be replaced with:
```python
def fixed_batch_process(im_data, model):
    batch_size = 512
    out = []
    for i in range(0, len(im_data), batch_size):
        batch = im_data[i:(i + batch_size)]
        out.append(model(batch))
    return tuple(torch.cat(v, dim=0) for v in zip(*out))
```

Reply:
Tim, your code is very compact and nice :).  Just use yours.

Reply:
Tim, I also added another commitment

https://github.com/timesler/facenet-pytorch/pull/105/commits/9e1879661ae272cbf885180829fe2a5f8a67b8f5
Please review it.

"perform nms inside the scale loop, rather than outside. This also help avoid GPU out of memory.
ideas similar to fixed_batch_process (restrict very large batch inside GPU)

Also 
change "image_inds = torch.cat(image_inds, dim=0).cpu()" to 
image_inds = torch.cat(image_inds, dim=0) and switch to cpu at the end of the function.